Background processing and Callback

xalakox opened the pull request:
We are adding a feature that allows sending actions (primarily because of product search timeouts) and a callback for results.

Both flags are not dependent on each other; any API operation in theory could be executed as a background job (right now only implemented on the product search).

If the backgroundJob flag is present, it will reply with a job Id that can be used to retrieve the results, or if used in conjunction, it would also use the callbackUrl to trigger a call with the results.

xalakox committed:
callback utility

xalakox committed:
background worker intercept on default route

xalakox committed:
easier to follow/debug process

xalakox committed:
WIP: callbackUrl call should be generic, here for testing

xalakox committed:
using the right attribute reference for background job creation

xalakox committed:
missing real param

xalakox committed:
wip: still not a successfull call

xalakox committed:
backgroundjob working

xalakox committed:
dont require outside of the job, to avoid memory leaks

xalakox committed:
move this to a generic callback handler

xalakox committed:
use a more generic approach to the handler

xalakox committed:
use the callbackUrl on the index

xalakox committed:
add callbackUrl abd backbroundjob parameter

xalakox committed:
backgroundJob test for productSearch

xalakox committed:
add a test for a api job type

xalakox committed:
add a callback job type test

xalakox committed:
no need for this change

mogii commented on controllers/__tests__/bookingsSearch-background.js:
```diff
@@ -0,0 +1,121 @@
+/* globals beforeAll describe it expect jest beforeEach */
```
Can you rename the file to be productsSearch-background.js?

mogii commented on index.js:
```diff
@@ -277,6 +279,44 @@ module.exports = async ({
+            });
+            return originalJson.call(this, data);
+          };
+        }
```
It seems that callbackUrl and backgroundJob can't be used at the same time. Is this intentional?

mogii replied:
I guess only the `type: api` backgroundJob cannot be used with callbackUrl. But within a controller, we can still add a job and pass a callbackUrl?

xalakox replied:
they can be used sim, then the background job finishes it will create a new callback job.

mogii commented on worker/index.js:
```diff
@@ -23,14 +22,88 @@ const worker = ({ plugins: pluginsParam }) => (id, disconnect) => {
+      });
+      return { success: true };
+    }
     console.log(`job ${jobId} > running ${pluginName}:${method} for ${userId}`);
```
remove console.log

xalakox replied:
this is a good idea since is useful when a job starts / ends; unlike celery that does this by default bull does not.

xalakox committed:
tests wait for the job to be completed

xalakox committed:
avoid console.error during callback tests

xalakox committed:
mock database operations to avoid connection persistance; [job tests]

mock database connections for the callback tests

xalakox committed:
close database connections after running every test

xalakox committed:
Revert "close database connections after running every test"

This reverts commit b767d714199e4c21d349efd695ab4aa88434dbf3.

xalakox committed:
remove un-neccesary console.log's

xalakox committed:
increase the timeout for the background job to complete

xalakox committed:
sinlence expected errors

xalakox committed:
test user / key / integration should be contained; do not use a new integration for subsequent caching tests

This allows tests to be ran in paralle
if we are creating the a new hint for every cache test we are not
testing the expected timeline for tests

xalakox committed:
use random test user for allotment tests only

xalakox committed:
not a valid dot notation for older javascript versions

xalakox committed:
running is not a valid bull status

xalakox committed:
user a new user and app for the bookingSearch [background]

xalakox committed:
file should be nanmed to what it actually is

xalakox committed:
was creating the user twice

xalakox committed:
don't try/catch jest errors

xalakox committed:
linting

xalakox committed:
do not mock the models

don't mock the model

xalakox committed:
don't clean the queue as other tests might use them

dont clean queue

xalakox committed:
from an instantiated server to supertest

uses a more efficient approach by directly test/run the Express app
without creating an actual HTTP server.

This should improve performance and reliability for API type jobs.

xalakox committed:
Revert "from an instantiated server to supertest"

This reverts commit c66f0e57cd52aef95584753595a7fdfe1e9344aa.

xalakox committed:
use the already available cache library, avoiding new insntances of the cache connection

xalakox committed:
emit a new type of event (bookingProductSearch:cache:save)

xalakox committed:
allow axios to be used directly on the plugins.

xalakox committed:
not url paramters

xalakox committed:
linting: is yaml not json

xalakox committed:
re-add missing references

xalakox committed:
maintance: we should also close any sequelize connections after running a set of tests

xalakox committed:
WIP: test is working, need to clean-up

xalakox committed:
remocing verbosity

xalakox committed:
silence worker console.logs
